src/speculators/data_generation/custom_worker.py

```diff
@@ -0,0 +1,115 @@
+"""Custom worker extension for hidden states capture."""
+
+import logging
+from itertools import islice
+from typing import Any
+
+import torch
+from vllm.distributed import get_pp_group, get_tp_group
+from vllm.model_executor.models.interfaces import supports_eagle3
+from vllm.sequence import IntermediateTensors
+
+__all__ = ["HiddenStatesWorkerExtension"]
+
+logger = logging.getLogger(__name__)
+
+
+class HiddenStatesWorkerExtension:
+    """Worker extension that adds hidden states capture functionality."""
+
+    _layer_ids: list[int]
+    _captured_states: list[Any] | None
+    _should_capture: bool
+    model_runner: Any
+
+    def _store_captured_states(self, aux_hidden_states):
+        if self._captured_states is None:
+            self._captured_states = aux_hidden_states
+        else:
+            for i, h in enumerate(aux_hidden_states):
+                self._captured_states[i] = torch.cat(
```

Comment:
Doing many cats might be costly. Consider keeping these at as a list, and concatting only when you need to read them (_get_captured_states)?